node/modules/lp2p/pubsub.go

```diff
@@ -31,8 +43,132 @@ func PubsubTracer() PubsubOpt {
 }
 
 func GossipSub(pubsubOptions ...PubsubOpt) interface{} {
-	return func(mctx helpers.MetricsCtx, lc fx.Lifecycle, host host.Host) (service *pubsub.PubSub, err error) {
-		return pubsub.NewGossipSub(helpers.LifecycleCtx(mctx, lc), host, paresOpts(host, pubsubOptions)...)
+	return func(mctx helpers.MetricsCtx, lc fx.Lifecycle, host host.Host, nn dtypes.NetworkName, bp dtypes.BootstrapPeers) (service *pubsub.PubSub, err error) {
+		bootstrappers := make(map[peer.ID]struct{})
+		for _, pi := range bp {
+			bootstrappers[pi.ID] = struct{}{}
+		}
+		_, isBootstrapNode := bootstrappers[host.ID()]
+
+		v11Options := []pubsub.Option{
+			// Gossipsubv1.1 configuration
+			pubsub.WithFloodPublish(true),
+			pubsub.WithPeerScore(
+				&pubsub.PeerScoreParams{
+					AppSpecificScore: func(p peer.ID) float64 {
+						// return a heavy positive score for bootstrappers so that we don't unilaterally prune
+						// them and accept PX from them.
+						// we don't do that in the bootstrappers themselves to avoid creating a closed mesh
+						// between them (however we might want to consider doing just that)
+						_, ok := bootstrappers[p]
+						if ok && !isBootstrapNode {
+							return 2500
+						}
+
+						// TODO: we want to  plug the application specific score to the node itself in order
+						//       to provide feedback to the pubsub system based on observed behaviour
+						return 0
+					},
+					AppSpecificWeight: 1,
+
+					// This sets the IP colocation threshold to 1 peer per
+					IPColocationFactorThreshold: 1,
+					IPColocationFactorWeight:    -100,
+					// TODO we want to whitelist IPv6 /64s that belong to datacenters etc
+					// IPColocationFactorWhitelist: map[string]struct{}{},
+
+					DecayInterval: time.Second,
+					DecayToZero:   0.01,
+
+					// this retains non-positive scores for 6 hours
+					RetainScore: 6 * time.Hour,
+
+					// topic parameters
+					Topics: map[string]*pubsub.TopicScoreParams{
+						build.BlocksTopic(nn): {
+							// expected 10 blocks/min
+							TopicWeight: 0.1, // max is 250, max mesh penalty is -10, single invalid message is -100
+
+							// 1 tick per second, maxes at 1 after 1 hour
+							TimeInMeshWeight:  0.00027, // ~1/3600
+							TimeInMeshQuantum: time.Second,
+							TimeInMeshCap:     1,
+
+							// deliveries decay after 1 hour, cap at 100 blocks
+							FirstMessageDeliveriesWeight: 25,      // max value is 2500
+							FirstMessageDeliveriesDecay:  0.99972, // 1 hour
```

Comment:
for things like this, could gossipsub provide helper functions? Being able to just say "this decays in one hour" is so much nicer than having to do the math yourself

Reply:
Yes indeed, I plan to write something along these lines but we are not there yet.

Reply:
I added and used a `pubsub.ScoreParameterDecay` which uses time to compute the decay factor.